core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionValidator.scala

```diff
@@ -0,0 +1,48 @@
+package org.broadinstitute.dsde.rawls.expressions
+
+import org.broadinstitute.dsde.rawls.RawlsExceptionWithErrorReport
+import org.broadinstitute.dsde.rawls.jobexec.MethodConfigResolver
+import org.broadinstitute.dsde.rawls.model.{AttributeString, ErrorReport, MethodConfiguration, ValidatedMethodConfiguration}
+import spray.http.StatusCodes
+
+import scala.util.Try
+
+object ExpressionValidator {
+  // inputsToParse is parameterized in order to validate in the presence or absence of the associated Method
+  // presence: inputs which are both empty and optional are pre-validated, so they are skipped here
+  // absence: validate all inputs normally
+
+  private[expressions] def validateAndParse(methodConfiguration: MethodConfiguration, inputsToParse: Map[String, AttributeString], parser: SlickExpressionParser) = {
+    val (emptyOutputs, outputsToParse) = methodConfiguration.outputs.partition { case (_, expr) => expr.value.isEmpty }
+
+    val parsed = ExpressionParser.parseMCExpressions(inputsToParse, outputsToParse, parser)
+
+    // empty output expressions are also valid
+    val validatedOutputs = emptyOutputs.keys.toSeq ++ parsed.validOutputs
+
+    ValidatedMethodConfiguration(methodConfiguration, parsed.validInputs, parsed.invalidInputs, validatedOutputs, parsed.invalidOutputs)
+  }
+
+  // validate a MC without retrieving the Method from the Repo: assume all inputs are required
+  def validateAndParseMCExpressions(methodConfiguration: MethodConfiguration, parser: SlickExpressionParser): ValidatedMethodConfiguration =
+    validateAndParse(methodConfiguration, methodConfiguration.inputs, parser)
+
+  // validate a MC, skipping optional empty inputs, and return failure when any inputs/outputs are invalid
+  def validateExpressionsForSubmission(methodConfiguration: MethodConfiguration, methodInputsToParse: Seq[MethodConfigResolver.MethodInput], emptyOptionalMethodInputs: Seq[MethodConfigResolver.MethodInput], parser: SlickExpressionParser): Try[ValidatedMethodConfiguration] = {
+    val inputsToParse = methodInputsToParse map { mi => (mi.workflowInput.fqn, AttributeString(mi.expression)) }
+
+    val validated = validateAndParse(methodConfiguration, inputsToParse.toMap, parser)
+
+    Try {
+      if (validated.invalidInputs.nonEmpty || validated.invalidOutputs.nonEmpty) {
+        val errors = s"Invalid inputs: ${validated.invalidInputs.mkString(",")} ; Invalid outputs: ${validated.invalidOutputs.mkString(",")}"
```

Comment:
Does this mean if I have 1 invalid input and 0 invalid outputs, my error string will look sort of like this:

`Invalid inputs: myinput ; Invalid outputs: `

Could we make it so that, in this sort of a scenario, it just reads like:

`Invalid inputs: myinput`

Reply:
oh ok I suppose :)